File: 2017/Day01.js
module.exports = function()
{
    const fs = require('fs');
    const readline = require('readline');

    const readInput = readline.createInterface({
        input: fs.createReadStream('Data/Day01.data')
    });

    let part = 1;
    
    console.log("----- part 1 -----");

    readInput
    .on('line', (line) => { 
        processLine(line);
    })
    .on('close', () => {
        process.exit(0);
    });

    function processLine(line)
    {
        let total = 0;
        let count = line.length;
        let offset= 1;

        if (line == "*")
        {
            console.log("----- part 2 -----");
            part = 2;
            return;
        }

        if (part == 2)
            offset = count/2;

        for (let i = 0; i < count; i++)
        {
            let c1 = line.charAt(i);
            let c2 = line.charAt((i+offset) % count);
            if (c1 == c2)
                total += +c1;
        }
        console.log("Total = " + total);
    }
}

File: 2017/C/Day01.c
#include <stdio.h>
#include <string.h>

void day1(void)
{
	int	  part = 1;
	FILE* file = fopen("Data/Day01.data", "r");
	
	if (file == NULL)
	{
		fprintf(stderr, "File not found!");
		return;
	}
	
	char buffer[1024];
	
	printf("---- PART 1 ----\n");
	
	while (! feof(file))
	{
		char* ptr = fgets(buffer, 1024, file);
		
		if (*ptr == '*')
		{
			part = 2;
			printf("---- PART 2 ----\n");
			continue;
		}
		
		int count = (int) strlen(ptr);
		int offset = part == 2 ? (count / 2) : 1;
		if (part == 2)
			offset = count/2;
		
		int total = 0;
		for (int i = 0; i < count; i++)
		{
			char c1 = ptr[i];
			char c2 = ptr[(i+offset) % count];
			if (c1 == c2)
				total += (c1 - '0');
		}

		printf("Total = %i\n", total);
	}
	fclose(file);
}

